metamist/parser/generic_parser.py

```diff
@@ -136,48 +186,178 @@ def process_fieldname(self, fieldname: str) -> str:
+        self.external_pid = external_pid
+
+        self.reported_sex = reported_sex
+        self.reported_gender = reported_gender
+        self.karyotype = karyotype
         self.meta = meta
 
+        self.samples: list[ParsedSample] = []
+
+    def to_sm(self) -> ParticipantUpsert:
+        """Convert to SM upsert model"""
+        samples = [s.to_sm() for s in self.samples]
+        return ParticipantUpsert(
+            external_id=self.external_pid,
+            reported_sex=self.reported_sex,
+            reported_gender=self.reported_gender,
+            karyotype=self.karyotype,
+            meta=self.meta,
+            samples=samples,
+        )
+
 
-class SampleMetaGroup:
+class ParsedSample:
     """Class for holding sample metadata grouped by id"""
 
     def __init__(
         self,
-        sample_id: str,
+        participant: ParsedParticipant | None,
         rows: GroupedRow,
+        internal_sid: str | None,
+        external_sid: str,
+        sample_type: str,
         meta: Dict[str, Any] = None,
     ):
-        self.id = sample_id
+        self.participant = participant
         self.rows = rows
+
+        self.internal_sid = internal_sid
+        self.external_sid = external_sid
+        self.sample_type = sample_type
         self.meta = meta
 
+        self.sequencing_groups: list[ParsedSequencingGroup] = []
+
+    def to_sm(self) -> SampleUpsert:
+        """Convert to SM upsert model"""
+        return SampleUpsert(
+            id=self.internal_sid,
+            external_id=self.external_sid,
+            type=self.sample_type,
+            meta=self.meta,
+            active=True,
+            sequencing_groups=[sg.to_sm() for sg in self.sequencing_groups],
+        )
+
 
-class SequenceMetaGroup:
+class ParsedSequencingGroup:
     """Class for holding sequence metadata grouped by type"""
 
     def __init__(
         self,
+        sample: ParsedSample,
+        rows: GroupedRow,
+        internal_seqgroup_id: int | None,
+        external_seqgroup_id: str | None,
+        sequence_type: str,
+        sequence_technology: str,
+        sequence_platform: str | None,
+        meta: dict[str, Any] | None,
+    ):
+        self.sample = sample
+        self.rows = rows
+
+        self.internal_seqgroup_id = internal_seqgroup_id
+        self.external_seqgroup_id = external_seqgroup_id
+        self.sequencing_type = sequence_type
+        self.sequencing_technology = sequence_technology
+        self.sequencing_platform = sequence_platform
+        self.meta = meta
+
+        self.assays: list[ParsedAssay] = []
+        self.analyses: list[ParsedAnalysis] = []
+
+    def to_sm(self) -> SequencingGroupUpsert:
+        """Convert to SM upsert model"""
+        return SequencingGroupUpsert(
+            type=self.sequencing_type,
+            technology=self.sequencing_technology,
+            platform=self.sequencing_platform,
+            meta=self.meta,
+            assays=[sq.to_sm() for sq in self.assays],
```

Comment:
Hm, I've tried testing locally and am getting `TypeError: 'NoneType' object is not iterable` at this stage.

Reply:
What were you running to get this? Could be worth adding as a test :)